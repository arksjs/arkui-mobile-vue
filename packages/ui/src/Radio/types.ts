import type {
  CheckCommonEmits,
  CheckCommonProps,
  CheckGroupCommonProps,
  ModelValue
} from '../Checkbox/types'

export interface RadioGroupProps extends CheckGroupCommonProps {
  modelValue?: ModelValue
}

export interface RadioGroupEmits {
  onChange?: (value: ModelValue) => void
}

export type RadioProps = CheckCommonProps

export type RadioEmits = CheckCommonEmits
