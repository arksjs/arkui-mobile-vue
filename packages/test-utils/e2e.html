<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
