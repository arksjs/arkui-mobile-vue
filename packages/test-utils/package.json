{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
