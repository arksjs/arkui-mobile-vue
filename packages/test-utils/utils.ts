export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
