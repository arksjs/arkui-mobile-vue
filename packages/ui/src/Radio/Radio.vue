<template>
  <label class="ak-radio" :class="classes" :style="styles">
    <input
      class="ak-radio_input ak-form-input"
      type="radio"
      :name="name2"
      :value="value"
      :disabled="disabled2"
      @change="onChange"
      ref="input"
    />
    <div class="ak-radio_box">
      <Icon class="ak-radio_icon" :icon="CircleOutlined" />
      <Icon class="ak-radio_checked-icon" :icon="CheckCircleFilled" />
      <span class="ak-radio_text" v-if="$slots.default">
        <slot></slot>
      </span>
    </div>
  </label>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { Icon } from '../Icon'
import { checkEmits, checkProps } from '../Checkbox/props'
import { useCheck } from '../Checkbox/use-check'
import CircleOutlined from '../Icon/icons/CircleOutlined'
import CheckCircleFilled from '../Icon/icons/CheckCircleFilled'

export default defineComponent({
  name: 'ak-radio',
  components: { Icon },
  props: {
    ...checkProps
  },
  emits: { ...checkEmits },
  setup(props, ctx) {
    return {
      ...useCheck(props, ctx, 'radio'),
      CircleOutlined,
      CheckCircleFilled
    }
  }
})
</script>
