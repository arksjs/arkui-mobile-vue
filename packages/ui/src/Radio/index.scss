@import '../style/var.scss';

.ta-radio {
  --ta-active-color: #{$primary-color};
  --ta-radio-icon-size: 20px;

  display: inline-flex;
  flex-direction: column;
  align-items: center;
  font-size: 17px;
  color: $title-color;
  text-align: left;
  position: relative;

  + .ta-radio {
    margin-left: 16px;
  }

  &_box {
    display: flex;
    width: 100%;
    height: 32px;
    box-sizing: border-box;
    align-items: center;
  }

  &_icon,
  &_checked-icon {
    --ta-icon-size: var(--ta-radio-icon-size);
    --ta-icon-color: #{$border-color};
  }

  &_checked-icon {
    display: none;

    --ta-icon-color: var(--ta-active-color);
  }

  &_input {
    position: absolute;
    left: 0;
    top: 0;
    opacity: 0;

    &:checked + .ta-radio_box .ta-radio_icon {
      display: none;
    }

    &:checked + .ta-radio_box .ta-radio_checked-icon {
      display: block;
    }
  }

  &_text {
    display: block;
    line-height: 1;
    margin-left: 12px;
  }

  &.disabled {
    opacity: 0.3;
  }
}

.ta-radio-group {
  display: flex;
  height: 32px;
  align-items: center;
  color: $title-color;
  box-sizing: border-box;
  flex: 1;

  &.vertical {
    flex-direction: column;
    height: auto;
    min-height: 32px;

    > .ta-radio {
      display: flex;
      width: 100%;

      + .ta-radio {
        margin-left: 0;

        &::before {
          content: '';
        }
      }
    }
  }
}
