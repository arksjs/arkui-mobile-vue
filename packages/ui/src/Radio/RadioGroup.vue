<template>
  <div class="ak-radio-group" :class="classes" ref="root">
    <slot>
      <Radio
        v-for="item in options2"
        :key="item.value"
        :checkedValue="item.value"
        >{{ item.label }}</Radio
      >
    </slot>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref } from 'vue'
import Radio from './Radio.vue'
import { useCheckGroup } from '../Checkbox/use-check'
import type { ModelValue } from '../Checkbox/types'
import { checkGroupProps } from '../Checkbox/props'
import { isStringNumberMix } from '../helpers/util'
import type { PropsToEmits } from '../helpers/types'
import type { RadioGroupEmits } from './types'

const isValue = (value: ModelValue) => isStringNumberMix(value)

export default defineComponent({
  name: 'ak-radio-group',
  components: { Radio },
  props: {
    ...checkGroupProps,
    modelValue: {
      type: [Number, String],
      validator: isValue,
      default: null
    }
  },
  emits: {
    'update:modelValue': isValue,
    change: isValue
  } as PropsToEmits<RadioGroupEmits>,
  setup(props, ctx) {
    const inputValue = ref<ModelValue>('')
    const { emit } = ctx

    const group = useCheckGroup(props, {
      name: 'radio',
      updateValue({ isChange, uid, children }) {
        let hasChecked = false
        let value: ModelValue = ''

        children.forEach(child => {
          const checked = uid ? uid === child.uid : child.getInputChecked()

          if (!hasChecked && checked) {
            hasChecked = true
            value = child.getValue()
            child.setChecked(true)
          } else {
            child.setChecked(false)
          }
        })

        inputValue.value = value

        if (isChange && inputValue.value !== props.modelValue) {
          emit('update:modelValue', value)
        }

        if (isChange) {
          emit('change', value)
        }

        return value
      },
      watchValue({ children, value }) {
        let hasChecked = false

        children.forEach(child => {
          if (!hasChecked && child.getValue() === value) {
            hasChecked = true
            inputValue.value = value
            child.setChecked(true)
          } else {
            child.setChecked(false)
          }
        })
      }
    })

    return {
      ...group
    }
  }
})
</script>
