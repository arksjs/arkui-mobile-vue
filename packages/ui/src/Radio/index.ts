import { multiWithInstall } from '../helpers/with-install'
import Radio from './Radio.vue'
import RadioGroup from './RadioGroup.vue'

export { Radio, RadioGroup }
export default Radio
export const install = multiWithInstall(Radio, [RadioGroup])
