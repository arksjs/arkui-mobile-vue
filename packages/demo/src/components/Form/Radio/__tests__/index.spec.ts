import { shallowMount } from '@vue/test-utils'
import { Radio, RadioGroup } from '@/Radio'

describe('Radio', () => {
  test('should render default correctly', () => {
    const wrapper = shallowMount(Radio, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop disabled=true correctly', () => {
    const wrapper = shallowMount(Radio, {
      props: {
        disabled: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('RadioGroup', () => {
  test('should render 2 Radio slot correctly', () => {
    const wrapper = shallowMount(RadioGroup, {
      slots: {
        default: [Radio, Radio]
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop disabled=true correctly', () => {
    const wrapper = shallowMount(RadioGroup, {
      props: {
        disabled: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
