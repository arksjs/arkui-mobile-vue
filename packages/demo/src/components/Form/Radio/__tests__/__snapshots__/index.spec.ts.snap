// Vitest Snapshot v1

exports[`Radio > should render default correctly 1`] = `
"<label class=\\"fx-radio fx-horizontal-hairline\\"><input class=\\"fx-radio_input fx-form-input\\" type=\\"radio\\" name=\\"\\" value=\\"\\">
  <div class=\\"fx-radio_box\\">
    <icon-stub icon=\\"[object Object]\\" spin=\\"false\\" size=\\"0\\" width=\\"0\\" height=\\"0\\" class=\\"fx-radio_icon\\"></icon-stub>
    <icon-stub icon=\\"[object Object]\\" spin=\\"false\\" size=\\"0\\" width=\\"0\\" height=\\"0\\" class=\\"fx-radio_checked-icon\\"></icon-stub>
    <!--v-if-->
  </div>
</label>"
`;

exports[`Radio > should render prop disabled=true correctly 1`] = `
"<label class=\\"fx-radio fx-horizontal-hairline disabled\\"><input class=\\"fx-radio_input fx-form-input\\" type=\\"radio\\" name=\\"\\" disabled=\\"\\" value=\\"\\">
  <div class=\\"fx-radio_box\\">
    <icon-stub icon=\\"[object Object]\\" spin=\\"false\\" size=\\"0\\" width=\\"0\\" height=\\"0\\" class=\\"fx-radio_icon\\"></icon-stub>
    <icon-stub icon=\\"[object Object]\\" spin=\\"false\\" size=\\"0\\" width=\\"0\\" height=\\"0\\" class=\\"fx-radio_checked-icon\\"></icon-stub>
    <!--v-if-->
  </div>
</label>"
`;

exports[`RadioGroup > should render 2 Radio slot correctly 1`] = `
"<div class=\\"fx-radio-group vertical\\">
  <fx-radio-stub checked=\\"false\\" disabled=\\"false\\" name=\\"\\" value=\\"\\"></fx-radio-stub>
  <fx-radio-stub checked=\\"false\\" disabled=\\"false\\" name=\\"\\" value=\\"\\"></fx-radio-stub>
</div>"
`;

exports[`RadioGroup > should render prop disabled=true correctly 1`] = `"<div class=\\"fx-radio-group vertical disabled\\"></div>"`;
